[1.4] [DRAFT] Unification and improvements in bootstrap scripts

DarioDaF opened the pull request:
### What is the bug?

#1859

### How did you fix the bug?

Part of it by bundling a BusyBox-like for windows and run all under bash script
Yet to write is to update Powershell by bundling a msi for it (or downloading it if necessary), but looking into regenerating env vars after install and deciding if it is ok installing a "system" update silently is ok...

### Are there alternatives to your fix?

Any compiled thing that needs to be our artifact is annoying, so if there is any decent no dep scripting language with a very small interpreter and flexible enough that could be an option

DarioDaF committed:
Busybox added and initial unification of install

DarioDaF committed:
Merge branch '1.4' into daf/1.4/uniform-bootstrap

DarioDaF committed:
Small change to show manual install only on Windows

DarioDaF committed:
Try to move stuff in the correct place and merge 1.4

Solxanich committed:
Started separating out component scripts

Solxanich committed:
The next big rework to move everything to be reorganized and trimmed.

Now it just needs to be functional :)

DarioDaF committed:
Restoring busybox and fixing some script code leaving debug strings

Solxanich committed:
Bug Fixes

- Fix start.bat quotes and exist
- fix Version parsing
- Fix error code error
- Fix Powershell invoke
- Fix manual install URL and use Curl/wget direct as backup

DarioDaF committed:
Further cleanup (removed hardcoded dotnet version) (NOT TESTED)

DarioDaF committed:
Fixing a dot

DarioDaF committed:
Fixing ifs and realpaths + WINDIR backslash for MonoMod (will not be needed in the future, but doesn't hurt)

DarioDaF committed:
Fixing quotes and adding chmod to ensure `start`s working linux

DarioDaF committed:
Moved version check on ScriptCaller to allow redirect of install to log

RunningDroid commented on patches/tModLoader/Terraria/release_extras/LaunchUtils/ScriptCaller.sh:
```diff
@@ -0,0 +1,88 @@
+	run_script ./Remove13_64Bit.sh 2>&1 | tee "$LogFile"
+fi
+
+run_script ./UnixLinkerFix.sh 2>&1 | tee -a "$LogFile"
```
Changes in a child process' environment don't get propagated to the parent process' environment, so this should probably be changed to source ./UnixLinkerFix.sh

RunningDroid commented on patches/tModLoader/Terraria/release_extras/LaunchUtils/ScriptCaller.sh:
```diff
@@ -0,0 +1,88 @@
+export install_dir
+export dotnet_dir
+export dotnet_version
+./InstallNetFramework.sh 2>&1 | tee -a "$LogFile"
```
This needs `run_script`

RunningDroid commented on patches/tModLoader/Terraria/release_extras/LaunchUtils/UnixLinkerFix.sh:
```diff
@@ -0,0 +1,27 @@
+# Provided for use in tModLoader deployment. 
+
+#chdir to path of the script and save it
+cd "$(dirname "$(realpath "$0")")"
```
If this script is sourced by ScriptCaller.sh then realpath errors out with "realpath: ./LaunchUtils/ScriptCaller.sh: No such file or directory" (because it's looking for LaunchUtils/LaunchUtils/ScriptCaller.sh.)
This can either be fixed by removing this line or by changing start-tModLoader.sh to use the absolute path when calling ScriptCaller.sh.

RunningDroid commented on patches/tModLoader/Terraria/release_extras/LaunchUtils/ScriptCaller.sh:
```diff
@@ -0,0 +1,88 @@
+echo "Verifying .NET...."
+echo "This may take a few moments."
+echo "Logging to $LogFile"
+
```
If you add `exec > >(tee "$LogFile") 2>&1` (redirect output to named pipe containing tee) here, then you don't need to have `2>&1 | tee "$LogFile"` after every command.

RunningDroid commented on patches/tModLoader/Terraria/release_extras/LaunchUtils/ScriptCaller.sh:
```diff
@@ -0,0 +1,88 @@
+./InstallNetFramework.sh 2>&1 | tee -a "$LogFile"
+
+echo "Attempting Launch..."
+sleep 1
```
This line appears to be unnecessary

RunningDroid commented on patches/tModLoader/Terraria/release_extras/LaunchUtils/ScriptCaller.sh:
```diff
@@ -0,0 +1,88 @@
+fi
+
+if [[ -f "$install_dir/dotnet" || -f "$install_dir/dotnet.exe" ]]; then
+	"$install_dir/dotnet" tModLoader.dll $*
```
`$*` here should probably be `"$@"` (including double quotes): https://github.com/koalaman/shellcheck/wiki/SC2048
 It'd also be nice if these `dotnet tModLoader.dll` lines were exec'd to clean up the process tree a little bit

RunningDroid commented on patches/tModLoader/Terraria/release_extras/start-tModLoader.sh:
```diff
@@ -1,22 +1,5 @@
-  dotnet tModLoader.dll $launch_args
-fi
+chmod +x ./LaunchUtils/ScriptCaller.sh
+./LaunchUtils/ScriptCaller.sh $*
```
My comments from the last lines of ScriptCaller.sh apply here too

RunningDroid commented on patches/tModLoader/Terraria/release_extras/start-tModLoaderServer.sh:
```diff
@@ -22,17 +20,5 @@ fi
-  dotnet tModLoader.dll $launch_args
-fi
+chmod +x ./LaunchUtils/ScriptCaller.sh
+./start-tModLoader.sh $launch_args
```
It'd be nice if this were exec'd too

RunningDroid commented on patches/tModLoader/Terraria/release_extras/LaunchUtils/BashUtils.sh:
```diff
@@ -0,0 +1,34 @@
+# Sourced from dotnet-install.sh
+# Check if a program is present or not
+machine_has() {
+  eval $invocation
```
`$invocation` doesn't appear to be defined anywhere?

RunningDroid commented on patches/tModLoader/Terraria/release_extras/LaunchUtils/BashUtils.sh:
```diff
@@ -0,0 +1,34 @@
+
+# Call a script setting its permission right for execution
+run_script() {
+  chmod +x $1
```
`$1` should be wrapped in double quotes in case the path to tModLoader's install directory has spaces.

RunningDroid commented on patches/tModLoader/Terraria/release_extras/LaunchUtils/InstallNetFramework.sh:
```diff
@@ -0,0 +1,65 @@
+#If the dotnet dir exists, we need to do some cleanup
+if [ -d "$dotnet_dir" ]; then
+  # Find all folders inside the dotnet dir that don't match our target version and nuke it
+  for folder in $(ls "$dotnet_dir"); do
```
[Iterating over ls output is fragile](https://github.com/koalaman/shellcheck/wiki/SC2045), use `for folder in "$dotnet_dir/*"; do` instead

Solxanich replied:
Tried this. 
Didn't work as expected - the any character didn't expand to search all items, and instead searched literal.

I'm comfortable keeping this fragility as the average person doesn't have reason to put stuff in there. 
"6.0.0" doesn't create issues afaik, and cleaning up old versions with same naming convention should be fine as well. 
For this one, if it ain't broke, and the fix is broke, don't spend time fixing the fix

RunningDroid commented on patches/tModLoader/Terraria/release_extras/LaunchUtils/InstallNetFramework.sh:
```diff
@@ -0,0 +1,65 @@
+#If the install directory for this specific dotnet version doesnt exist, grab the installer script and run it.
+echo "Checking for Updates to .NET"
+if [ ! -d "$install_dir" ]; then
+  installLogs="LaunchLogs/install.log"
```
installLogs doesn't appear to be used anywhere

Solxanich committed:
Remove RealPath as its unnecessary

Move export to be at the variable assignment level to avoid chopping

RunningDroid reviewed (commented):
I did some testing and found some things that could be fixed; Note that the comments about exec'ing are mostly aesthetic since the extra processes don't eat any CPU and only consume ~32.5 KB of memory (on my machine at least.)

Solxanich committed:
Fix missing Run_Script function call

Solxanich committed:
Fix dotnet install.sh invoke

Solxanich committed:
Mac Testing Completed

Fix requiring to source UnixLinkerFix
Fix Log File appending behaviour to not grow infinitely

Solxanich committed:
Better Logging update

Solxanich commented:
Thanks RunningDroid! That's great you took the time to test and sort out problems.
It's unfortunate I missed your review as I wasn;t subscribed to notifications on this, but glad to see I implemented some of it already. 
I'll start work on implementing the remainder soon

Solxanich committed:
Logging Improvements

Also some other things.

Solxanich committed:
M1 launch fix

Shockah commented on patches/tModLoader/Terraria/release_extras/LaunchUtils/InstallNetFramework.sh:
```diff
@@ -0,0 +1,67 @@
+    mkdir "$dotnet_dir"
+
+    echo "It has been detected that your system failed to install the dotnet portables automatically. Will now proceed manually."
+	echo "Manual installation can take up to an additional 5 minutes"
```
Weird indent

Shockah reviewed (commented):
Maybe the indents should be standardized? In some places they are 2 spaces, in other they are 1 tab.

DarioDaF committed:
Only tabs/whitespaces and unused stuff fixes

DarioDaF committed:
Exported WINDIR (export or not doesn't do anything on win, but conceptually this is a propagated change)

DarioDaF committed:
Prioritize direct link install on win and allow for archive to be already present + small tab fix

Chicken-Bones commented on patches/tModLoader/Terraria/release_extras/LaunchUtils/InstallNetFramework.sh:
```diff
@@ -33,35 +36,43 @@ echo "Cleanup Complete"
+
+		# Allow for zip to be already delivered by steam win on system
+		# and placed in the root directory with this name convention:
+		dotnet_portable_archive="$root_dir/win-x64-$dotnet_version.zip"
```
I'd use the full `dotnet-runtime-$dotnet_version-win-x64.zip` so anyone looking at it knows what it is. Also, maybe only delete it if it was auto-downloaded? Otherwise steam is likely to try and restore it inconsistently.

Chicken-Bones commented on patches/tModLoader/Terraria/release_extras/LaunchUtils/InstallNetFramework.sh:
```diff
@@ -33,35 +36,43 @@ echo "Cleanup Complete"
+			echo "Downloading win x64 portable dotnet runtime..."
+			file_download "$dotnet_portable_archive" "https://dotnetcli.azureedge.net/dotnet/Runtime/$dotnet_version/dotnet-runtime-$dotnet_version-win-x64.zip"
+		else
+			echo "Runtime archive already present"
```
Message should say, `Found dotnet-runtime-$dotnet_version-win-x64.zip`

Chicken-Bones commented on patches/tModLoader/Terraria/release_extras/LaunchUtils/InstallNetFramework.sh:
```diff
@@ -33,35 +36,43 @@ echo "Cleanup Complete"
 #If the install directory for this specific dotnet version doesnt exist, grab the installer script and run it.
 echo "Checking for Updates to .NET"
```
Message should say `Checking dotnet install...`

Chicken-Bones commented on patches/tModLoader/Terraria/release_extras/LaunchUtils/InstallNetFramework.sh:
```diff
@@ -33,35 +36,43 @@ echo "Cleanup Complete"
 		run_script ./dotnet-install.sh --channel "$channel" --install-dir "$install_dir" --runtime "$dotnet_runtime" --version "$dotnet_version"
 	fi
 fi
 echo "Finished Checking for Updates"
```
Message should say `dotnet present in $install_dir`

Chicken-Bones reviewed (commented):
Few minor comments

Chicken-Bones commented on patches/tModLoader/Terraria/release_extras/LaunchUtils/ScriptCaller.sh:
```diff
@@ -0,0 +1,90 @@
+fi
+
+if [[ -f "$install_dir/dotnet" || -f "$install_dir/dotnet.exe" ]]; then
+	echo "Launched Using Local Dotnet" 2>&1 | tee -a "$LogFile"
+	chmod a+x "$install_dir/dotnet"
+	exec "$install_dir/dotnet" tModLoader.dll "$@"
+else
+	echo "Launched Using System Dotnet" 2>&1 | tee -a "$LogFile"
```
Maybe change the tense and case of these
`Launching using local dotnet...`
`Launching using system dotnet...`

DarioDaF committed:
Text messages minor fixes

DarioDaF committed:
Order now is zip if w64 and already present, ps install and then manual only if w64

Solxanich committed:
Massive Cleanup of Install Net Framework Script

Solxanich committed:
Bug Fixes

Solxanich committed:
Add Win32 platform Exception

Whether it forwards to SDL isn't an issue.
Just attempting failing early goes a long ways as the log catches it

Solxanich committed:
Export Path on windows for safety sake
